### PR TITLE
Adding the QuantileSplitter

### PR DESCRIPTION
**Pull request by florisvdf**
# Changes

Adds the `QuantileSplitter` and a new fixture for testing splits: `dataset_with_realistic_assays`.

The `QuantileSplitter` is intended to reserve a fraction of hit variants for the test set. Hit variants are defined as variants with target values that exceed the target values of all samples in the training set. The threshold for classifying variants as "hit variant" is determined by the `quantile` parameter.  

The `QuantileSplitter` divides a dataset into a lower, and an upper target quantile. The `fraction` parameter then determines what fraction of the lower quantile will make up the training data. The test data is then composed by uniformly sampling `1 - fraction` from the lower quantile and `1 - fraction` from the upper quantile. 

It's important to note that since the quantile split is dependent on the target values, the split is currently only defined for a single target. Additionally, this version currently only splits the data into a train and test set. There is no support yet for additional sets.

**Points of discussion**

1. I've noticed that all previously implemented splits share the same signature. Is that something that we want to enforce? If so, it will require a bit of refactoring of the `QuantileSplitter`. Perhaps then also warrants addition of an ABC `Splitter` class.
2. To properly test the `QuantileSplitter`, I've created a new fixture `dataset_with_realistic_assays`. The previous dataset fixtures with assay data have mostly constant target values. Is there a reason for this? If not, should we update them to have varied target distributions?

# Checklist

- [x] I broke the PR down so that it contains a reasonable amount of changes for an effective review
- [x] I performed a self-review of my code. Amongst other things, I have commented my code in hard-to-understand areas.
- [ ] I made corresponding changes to the documentation
- [x] I added tests that prove my fix is effective or that my feature works
- [ ] I accounted for dependent changes to be merged and published in downstream modules

### CODE REVIEW

File: src/proteingym/base/splits.py [hunk 1]
```diff
@@ -95,6 +96,35 @@ def _sequences_to_mask(selection: list[Seq], *, all_sequences: list[Seq]) -> lis
     return mask
 
 
+def _subsample_mask(
+    mask: npt.NDArray,
+    fraction: float,
+    random_state: int | np.random.RandomState | None = None,
+) -> npt.NDArray:
+    """
```
Comment (hredestig) [inline src/proteingym/base/splits.py:104]: Nitpick (we should have a linter for this) - docstrings are
```python
"""Title / Description"""
```
or if longer like in this case
```python
"""Title

Long description
"""
```

File: src/proteingym/base/splits.py [hunk 2]
```diff
@@ -95,6 +96,35 @@ def _sequences_to_mask(selection: list[Seq], *, all_sequences: list[Seq]) -> lis
...
+    Subsample True values in a boolean mask.
+
+    args
```
Comment (hredestig) [inline src/proteingym/base/splits.py]: Args:

File: src/proteingym/base/splits.py [hunk 3]
```diff
@@ -146,6 +176,83 @@ def _unique_sequences_for_targets(
     return list(sequences)
 
 
+class QuantileSplitter:
+    """Split a dataset a random training set and a test set with a sample of hit
```
Comment (hredestig) [inline src/proteingym/base/splits.py]: docstring format like above

File: src/proteingym/base/splits.py [hunk 4]
```diff
@@ -146,6 +176,83 @@ def _unique_sequences_for_targets(
...
+    variants defined as values exceeding a quantile threshold. This split is only
+    defined for a single target in the dataset, because the quantile threshold can only
+    be defined for a single target.
+    Args:
```
Comment (hredestig) [inline src/proteingym/base/splits.py:191]: empty line

File: src/proteingym/base/splits.py [hunk 5]
```diff
@@ -146,6 +176,83 @@ def _unique_sequences_for_targets(
...
+
+        Returns:
+            Subsets: The subsets containing the splits.
+        """
+        train_assay_slices = []
+        test_assay_slices = []
+        for assay in dataset.assays:
+            target_names_in_assay = [e.name for e in assay.fields]
+            if target not in target_names_in_assay:
+                columns = []
```
Comment (hredestig) [inline src/proteingym/base/splits.py]: pointless assignment, just 
```python
                train_assay_slices.append(AssaySlice(records=None, columns=[]))
```

File: src/proteingym/base/splits.py [hunk 6]
```diff
@@ -146,6 +176,83 @@ def _unique_sequences_for_targets(
...
+                train_assay_slices.append(AssaySlice(records=None, columns=columns))
+                test_assay_slices.append(AssaySlice(records=None, columns=columns))
+            else:
+                columns = [assay.sequence_feature_name, target]
+                df = assay.to_df(target_names=[target])
```
Comment (hredestig) [inline src/proteingym/base/splits.py]: slightly more efficient would be  something like
```python
target_index = assay.fields.index(target)
np.array(r[target_index] for r in assay.records)
```
since you would avoid to create the dataframe which is then never used

File: tests/conftest.py [hunk 7]
```diff
@@ -210,6 +210,74 @@ def dataset_with_assays() -> Dataset:
...
+        )
+        for i, s in enumerate(
+            ["AA", "CC", "DD", "EE", "FF", "GG", "HH", "II", "JJ", "KK"]
+        )
+    ]
+    assay1 = Assay(
+        name="assay2",
+        records=[
+            (sequences[0], 0.12, 1.83),
+            (sequences[0], 0.47, 1.21),  # duplicate sequence to handle that
```
Comment (hredestig) [inline tests/conftest.py:231]: makes sense - although at some point we agreed that in `Assay` we don't have duplicate sequences. It should be a pre-processing responsibility to from `AssayRaw` with duplicates to `Assay` with summarized data.

File: tests/conftest.py [hunk 8]
```diff
@@ -210,6 +210,74 @@ def dataset_with_assays() -> Dataset:
     return dataset
 
 
+@pytest.fixture
+def dataset_with_realistic_assays() -> Dataset:
```
Comment (hredestig) [inline tests/conftest.py]: realistic is maybe a big word here :) `dataset_with_varying_targets`

**Review by hredestig — CHANGES_REQUESTED** [review]
Looks good, just some minor nitpicks

File: tests/conftest.py (hunk 7 above)
Comment (florisvdf) [reply tests/conftest.py:231]: Ok. I have left this unchanged for now to mirror `dataset_with_assays`.

**Review by hredestig — APPROVED** [review]
